sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala

```diff
@@ -351,6 +351,41 @@ abstract class SchemaPruningSuite
     }
   }
 
+  testSchemaPruning("SPARK-34638: nested column prune on generator output") {
+    val query1 = spark.table("contacts")
+      .select(explode(col("friends")).as("friend"))
```

Comment:
Can this work correctly if `explode` has another func in its child, e.g., `explode(filter(friends, x -> x is not null))`?

Reply:
Let me add a test to verify it.

Reply:
Oh, because in the lambda function you access `friends`, Spark cannot prune it.

Reply:
The higher order function expression is a special case. I just looked at it with more details. For example, `ArrayFilter` holds the complex expression `friends`, so from the point of view of Spark, `friends` is thought to be used by `ArrayFilter`.

I think it has some space to optimize it more, targeting higher order function expression. It is out of the scope of this PR, however. I will take some time to look at this after.